### DISCUSSION TITLE
make `withLogsRecorder` a higher order function

### DISCUSSION THREAD

**Discussion by ryanleecode** [polkadot-api/polkadot-api#683] [General]
The current api of `withLogsRecorder` isn't fully pipeable and requires an annoymous function to move the JSONRpcProvider to second parameter.

For example

```ts
pipe(
  getWsProvider(endpoint),
  parsed(fixUnorderedBlocks),
  (provider) => withLogsRecorder(console.trace, provider),
  polkadotApi.createClient
)
```

it would be better if the function signature was

```ts
declare const withLogsRecorder: (persistLog: (log: string) => void) => (input: JsonRpcProvider) => JsonRpcProvider;
```

then you can do this

```ts
pipe(
  getWsProvider(endpoint),
  parsed(fixUnorderedBlocks),
  withLogsRecorder(console.trace),
  polkadotApi.createClient
)
```

this would obviously be a **breaking change** but, it is suggestion for next major release.

**Comment by josepot** [comment]
It doesn't really have to be a breaking change if we make the function curried... but I really don't see that much value into it, TBH.

**Comment by voliva** [comment]
Functional programming libraries often expose a curry function to do exactly that. [Ramda ](https://ramdajs.com/docs/#curry) has it, for instance.

Wouldn't that work?

**Comment by josepot** [comment]
I mean... The more I use the `withLogsRecorder`, the more I regret not making it a higher order function from the get go 🙈 @ryanleecode has a point!
